File: skills/code-architect/references/code-examples/swift/Visitor.swift
// Visitor — separate algorithms from objects via accept/visit (double dispatch).

protocol Visitor {
    func visit(_ node: NumberNode)
    func visit(_ node: TextNode)
}

protocol Element {
    func accept(_ visitor: Visitor)
}

struct NumberNode: Element {
    let value: Double
    func accept(_ visitor: Visitor) { visitor.visit(self) }
}

struct TextNode: Element {
    let value: String
    func accept(_ visitor: Visitor) { visitor.visit(self) }
}

// A concrete visitor implements one algorithm over all element types.
final class DescribeVisitor: Visitor {
    private(set) var output: [String] = []

    func visit(_ node: NumberNode) {
        output.append("number(\(node.value))")
    }

    func visit(_ node: TextNode) {
        output.append("text(\"\(node.value)\")")
    }
}

func runDemo() {
    let elements: [Element] = [
        NumberNode(value: 3.14),
        TextNode(value: "hi"),
        NumberNode(value: 42)
    ]

    let describer = DescribeVisitor()
    elements.forEach { $0.accept(describer) }
    describer.output.forEach { print($0) }
}

runDemo()

File: skills/code-architect/references/code-examples/go/visitor.go
// Visitor: separate algorithms from the object structure they operate on,
// using double dispatch via accept/visit.
package main

import "fmt"

// Visitor declares one visit method per concrete element type.
type Visitor interface {
	VisitCircle(c circle) string
	VisitSquare(s square) string
}

// Element accepts a visitor and routes to the right visit method.
type Element interface {
	Accept(v Visitor) string
}

type circle struct{ radius int }

func (c circle) Accept(v Visitor) string { return v.VisitCircle(c) }

type square struct{ side int }

func (s square) Accept(v Visitor) string { return v.VisitSquare(s) }

// areaVisitor is one algorithm defined entirely outside the elements.
type areaVisitor struct{}

func (areaVisitor) VisitCircle(c circle) string {
	return fmt.Sprintf("circle area ~= %d", 3*c.radius*c.radius)
}

func (areaVisitor) VisitSquare(s square) string {
	return fmt.Sprintf("square area = %d", s.side*s.side)
}

func main() {
	elements := []Element{circle{radius: 2}, square{side: 3}}
	v := areaVisitor{}
	for _, e := range elements {
		fmt.Println(e.Accept(v))
	}
}
